### PR TITLE
Add Virtualization to Redoc

### PR DESCRIPTION
**Pull request by audipasuatmadi**
## Unit Tests
image
image
image

## Demo Video
Sent on MS Teams.

### CODE REVIEW

File: src/components/Virtualization/VirtualizedContent.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    virtualizer.scrollToIndex(idx, {
+      align: 'start',
+    });
+  }, [selectedTag]);
+
+  return (
+    <div ref={scrollableRef} style={{ height: '100dvh', width: '100%', overflowY: 'auto' }}>
+      <div
+        style={{
+          height: virtualizer.getTotalSize() + 257,
```
Comment (alfinm01) [inline src/components/Virtualization/VirtualizedContent.tsx]: What is this specific number means?
Will this cause an issue in the web's dimension responsiveness? i.e. if we open the API docs in smaller devices

File: src/components/Virtualization/VirtualizedContent.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+          <div
+            key={virtualItem.key}
+            data-index={virtualItem.index}
+            ref={virtualizer.measureElement}
+            style={{
+              position: 'absolute',
+              top: 0,
+              left: 0,
+              width: '100%',
+              transform: `translateY(${virtualItem.start + 500}px)`,
```
Comment (alfinm01) [inline src/components/Virtualization/VirtualizedContent.tsx]: Same as my comment above.
Will the `+ 500` cause an issue in the web's dimension responsiveness? i.e. if we open the API docs in smaller devices

**Review by NGanip — APPROVED** [review]
Nice changes

**Comment by alfinm01** [comment]
Yup beautiful indeed

**Comment by firdausunit4** [comment]
Nice work!

**Review by firdausunit4 — APPROVED** [review]

File: src/components/Virtualization/VirtualizedContent.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            }}
+          >
+            <ContentItem
+              key={renderables[virtualItem.index].id}
+              item={renderables[virtualItem.index] as any}
```
Comment (kandaliakiki) [inline src/components/Virtualization/VirtualizedContent.tsx]: can we use ContentItemModel instead of any ? for better readibilty

**Comment by audipasuatmadi** [comment]
> Nice changes



> Yup beautiful indeed



> Nice work!

Thank you.

File: src/components/Virtualization/VirtualizedContent.tsx (hunk 3 above)
Comment (audipasuatmadi) [reply src/components/Virtualization/VirtualizedContent.tsx]: Good point, thank you. Will work on it.

Comment (audipasuatmadi) [reply src/components/Virtualization/VirtualizedContent.tsx]: Anything else? I'll push it in one-go.

File: src/components/Virtualization/VirtualizedContent.tsx (hunk 1 above)
Comment (kandaliakiki) [inline src/components/Virtualization/VirtualizedContent.tsx]: after discussion with audi , i think this +257 is not needed

File: src/components/Virtualization/VirtualizedContent.tsx (hunk 2 above)
Comment (kandaliakiki) [inline src/components/Virtualization/VirtualizedContent.tsx]: after discussion with audi , i think this +500px is not needed

File: src/components/__tests__/useItemReverseIndex.test.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      { id: 'item1', name: 'Item 1' },
+      { id: 'item2', name: 'Item 2' },
+      { id: 'item3', name: 'Item 3' },
+    ];
+
+    let result: MenuItemReverseIndexToVirtualIndex | undefined;
+
+    // Hooks can only be used inside a React component.
+    // This component is just to host the hook.
+    const TestComponent = ({ items }: { items: IMenuItem[] }) => {
```
Comment (kandaliakiki) [inline src/components/__tests__/useItemReverseIndex.test.tsx:37]: minor ,  the TestComponent shows up a lot in the tests. Could be a spot to apply DRY , but I leave it up to you to decide

File: src/components/__tests__/useItemReverseIndex.test.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    const newItems: any[] = [
+      { id: 'newItem1', description: 'New Item 1' },
+      { id: 'newItem2', description: 'New Item 2' },
+      { id: 'newItem3', description: 'New Item 3' },
+    ];
+
+    // Hooks can only be used inside a React component.
+    // This component is just to host the hook.
+    const TestComponent = ({ items }: { items: IMenuItem[] }) => {
```
Comment (kandaliakiki) [inline src/components/__tests__/useItemReverseIndex.test.tsx:76]: minor , the TestComponent shows up a lot in the tests. Could be a spot to apply DRY , but I leave it up to you to decide

File: src/components/__tests__/useItemReverseIndex.test.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      const { reverseIndexToVirtualIndex } = useItemReverseIndex(items);
+      result = reverseIndexToVirtualIndex;
+      return null;
+    };
```
Comment (kandaliakiki) [inline src/components/__tests__/useItemReverseIndex.test.tsx:80]: minor , the TestComponent shows up a lot in the tests. Could be a spot to apply DRY , but I leave it up to you to decide

File: src/components/__tests__/useItemReverseIndex.test.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      { id: 'item2', description: 'Item 2' },
+    ];
+
+    const newItems: any[] = [
+      { id: 'newItem1', description: 'New Item 1' },
+      { id: 'newItem2', description: 'New Item 2' },
+      { id: 'newItem3', description: 'New Item 3' },
+    ];
+
+    // Hooks can only be used inside a React component.
```
Comment (kandaliakiki) [inline src/components/__tests__/useItemReverseIndex.test.tsx:74]: minor , the TestComponent shows up a lot in the tests. Could be a spot to apply DRY , but I leave it up to you to decide

File: src/components/__tests__/useItemReverseIndex.test.tsx (hunk 5 above)
Comment (kandaliakiki) [inline src/components/__tests__/useItemReverseIndex.test.tsx:76]: minor , the TestComponent shows up a lot in the tests. Could be a spot to apply DRY , but I leave it up to you to decide

File: src/components/Virtualization/VirtualizedContent.tsx (hunk 1 above)
Comment (audipasuatmadi) [reply src/components/Virtualization/VirtualizedContent.tsx]: Thank you for the good discussion, code changes pushed.

File: src/components/Virtualization/VirtualizedContent.tsx (hunk 2 above)
Comment (audipasuatmadi) [reply src/components/Virtualization/VirtualizedContent.tsx]: Thank you for the good discussion, code changes pushed.

File: src/components/__tests__/useItemReverseIndex.test.tsx (hunk 5 above)
Comment (audipasuatmadi) [reply src/components/__tests__/useItemReverseIndex.test.tsx:76]: Thank you, valid points. It's just I'm afraid we cannot do that as we need the values of the hook from the TestComponent. I'd argue perhaps we'll keep it `KISS` (Keep it simple and stupid) at this stage.

Any thoughts?

Comment (kandaliakiki) [reply src/components/__tests__/useItemReverseIndex.test.tsx:76]: agreed

**Comment by audipasuatmadi** [comment]
Latest results:
image
image

**Comment by kandaliakiki** [comment]
Brilliant work !

**Comment by alfinm01** [comment]
Sorry, just realized my comment is on pending review, not submitted yet lmao
Great work Di

**Comment by audipasuatmadi** [comment]
Thanks Alfin. Yes it was the dirts I haven't cleaned up during the development (did some trial and error).
Now it has been cleaned.

**Pull request merged**